tests/DeadLetterQueueTest.cc

```diff
@@ -0,0 +1,390 @@
+    ConsumerConfiguration consumerConfig;
+    consumerConfig.setDeadLetterPolicy(dlqPolicy);
+    consumerConfig.setNegativeAckRedeliveryDelayMs(100);
+    consumerConfig.setConsumerType(consumerType_);
+    Consumer consumer;
+    ASSERT_EQ(ResultOk, client.subscribe(topic, subName, consumerConfig, consumer));
+
+    Consumer deadLetterQueueConsumer;
+    ConsumerConfiguration dlqConsumerConfig;
+    dlqConsumerConfig.setSubscriptionInitialPosition(InitialPositionEarliest);
+    ASSERT_EQ(ResultOk, client.subscribe(dlqTopic, "dlq-sub", dlqConsumerConfig, deadLetterQueueConsumer));
+
+    Producer producer;
+    ASSERT_EQ(ResultOk, client.createProducer(topic, producerConf_, producer));
+
+    const int num = 10;
+    Message msg;
+    for (int i = 0; i < num; ++i) {
+        msg = MessageBuilder().setContent(std::to_string(i)).build();
+        ASSERT_EQ(ResultOk, producer.send(msg));
+    }
+
+    // nack all msg.
+    for (int i = 0; i < dlqPolicy.getMaxRedeliverCount() * num + num; ++i) {
+        ASSERT_EQ(ResultOk, consumer.receive(msg));
+        consumer.negativeAcknowledge(msg);
+    }
+
+    // Use this subscription to ensure that messages are sent to the DLQ.
+    for (int i = 0; i < num; i++) {
+        ASSERT_EQ(ResultOk, deadLetterQueueConsumer.receive(msg));
+        ASSERT_TRUE(!msg.getDataAsString().empty());
+        ASSERT_TRUE(msg.getProperty(SYSTEM_PROPERTY_REAL_TOPIC).find(topic));
+        ASSERT_TRUE(!msg.getProperty(PROPERTY_ORIGIN_MESSAGE_ID).empty());
+    }
+
+    // If there is no initial subscription, then the subscription will not receive the DLQ messages sent
+    // before the subscription.
+    Consumer initDLQConsumer;
+    ConsumerConfiguration initDLQConsumerConfig;
+    dlqConsumerConfig.setSubscriptionInitialPosition(InitialPositionLatest);
+    ASSERT_EQ(ResultOk, client.subscribe(dlqTopic, dlqInitSub, initDLQConsumerConfig, initDLQConsumer));
+    for (int i = 0; i < num; i++) {
+        ASSERT_EQ(ResultOk, initDLQConsumer.receive(msg, 1000));
+        ASSERT_TRUE(!msg.getDataAsString().empty());
+        ASSERT_TRUE(msg.getProperty(SYSTEM_PROPERTY_REAL_TOPIC).find(topic));
+        ASSERT_TRUE(!msg.getProperty(PROPERTY_ORIGIN_MESSAGE_ID).empty());
+    }
+    ASSERT_EQ(ResultTimeout, initDLQConsumer.receive(msg, 200));
+}
+
+bool isBatchs[2] = {true, false};
+bool isMultiTopics[2] = {true, false};
+ConsumerType subTypes[2] = {ConsumerType::ConsumerShared, ConsumerType::ConsumerKeyShared};
+
+std::vector<std::tuple<bool, bool, ConsumerType>> getValues() {
+    std::vector<std::tuple<bool, bool, ConsumerType>> values;
+    for (const auto isBatch : isBatchs) {
+        for (const auto isMultiTopic : isMultiTopics) {
+            for (const auto subType : subTypes) {
+                values.emplace_back(std::make_tuple(isBatch, isMultiTopic, subType));
+            }
+        }
+    }
+    return values;
+}
+
+INSTANTIATE_TEST_CASE_P(Pulsar, DeadLetterQueueTest, ::testing::ValuesIn(getValues()));
```

Comment:
```suggestion
INSTANTIATE_TEST_SUITE_P(Pulsar, DeadLetterQueueTest,
                         testing::Combine(testing::Values(true, false), testing::Values(true, false),
                                          testing::Values(ConsumerType::ConsumerShared,
                                                          ConsumerType::ConsumerKeyShared)),
                         [](const testing::TestParamInfo<DeadLetterQueueTest::ParamType> &info) {
                             return "isBatch_" + std::to_string(std::get<0>(info.param)) + "_isMultiTopics_" +
                                    std::to_string(std::get<1>(info.param)) + "_subType_" +
                                    std::to_string(std::get<2>(info.param));
                         });
```

P.S. The last argument here is optional. Without it, the tests will be:

```
Pulsar/DeadLetterQueueTest.testInitSubscription/0
Pulsar/DeadLetterQueueTest.testInitSubscription/1
...
Pulsar/DeadLetterQueueTest.testInitSubscription/7
```

With it, the tests wil be:

```
Pulsar/DeadLetterQueueTest.testInitSubscription/isBatch_1_isMultiTopics_1_subType_1
Pulsar/DeadLetterQueueTest.testInitSubscription/isBatch_1_isMultiTopics_1_subType_3
Pulsar/DeadLetterQueueTest.testInitSubscription/isBatch_1_isMultiTopics_0_subType_1
Pulsar/DeadLetterQueueTest.testInitSubscription/isBatch_1_isMultiTopics_0_subType_3
Pulsar/DeadLetterQueueTest.testInitSubscription/isBatch_0_isMultiTopics_1_subType_1
Pulsar/DeadLetterQueueTest.testInitSubscription/isBatch_0_isMultiTopics_1_subType_3
Pulsar/DeadLetterQueueTest.testInitSubscription/isBatch_0_isMultiTopics_0_subType_1
Pulsar/DeadLetterQueueTest.testInitSubscription/isBatch_0_isMultiTopics_0_subType_3
```
```